### PR TITLE
Fix sh script (de)activate

### PR DESCRIPTION
**Pull request by x0id**
# Description

The POSIX shell kerl activate/deactivate code has been improved and refactored to fix environment setup/cleanup issues (https://github.com/kerl/kerl/issues/458).

The test suite for POSIX shell based activate/deactivate has been improved.

This PR partially (for the POSIX shell) resolves the issue https://github.com/kerl/kerl/issues/458.

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/kerl/kerl/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
👍 this is a very big change in only two commits that don't have a thorough description. I fear it'll make for a very hard review.

> improved

What was improved?

> refactored to fix environment setup/cleanup issues

What were the detected issues?

---

Tests are passing, and that's fine, but I'd like to be aware of what were actually achieving here. Thanks.

**Comment by x0id** [comment]
There were actually no issues in setting up/cleaning up the POSIX shell environment. But the annoying (me) fact is that a bunch of internal variables are present in the working environment, for example:
```
_KERL_ERL_AFLAGS_SET=
_KERL_ERL_CALL_REMOVABLE=/Users/user/kerl/25.2/lib/erl_interface-5.3/bin
_KERL_MANPATH_REMOVABLE=/Users/user/kerl/25.2/lib/erlang/man:/Users/user/kerl/25.2/man
_KERL_MANPATH_SET=x
_KERL_PATH_REMOVABLE=/Users/user/kerl/25.2/bin
_KERL_REBAR_CACHE_DIR_SET=
_KERL_REBAR_PLT_DIR_SET=
_KERL_SAVED_REBAR_CACHE_DIR=
_KERL_SAVED_REBAR_PLT_DIR=
```
and this could be easily avoided.

The PR logically consists of two updates:

1. Refactoring of the activate/cleanup code
2. Expanding test by adding more edge cases

The refactoring is based on a functional programming approach. Instead of creating the `kerl_deactivate` function which checks for various internal environment variables to see what exactly needs to be reverted, we generate the `kerl_deactivate` code which stores internally all the cleanup steps.

Instead of
```
SAVE_A=A
A="new value"
```
and
```
if [ -n "$SAVE_A" ]; then
    A=$SAVE_A
    unset SAVE_A
fi
```
we do
```
CLEANUP="$CLEANUP; A=$A"
eval "
function cleanup() {
    $CLEANUP
    unset -f cleanup
}
"
unset CLEANUP
```
and later to clean up environment we simply call this function:
```
cleaup
```

Real example of generating cleanup code:
```
$ ./kerl emit-activate 25.2 25.2-build /Users/user/kerl/25.2 >/tmp/activate.sh
$ vi /tmp/activate.sh
$ . /tmp/activate.sh
$ type -a kerl_deactivate
kerl_deactivate is a function
kerl_deactivate () 
{ 
    set -o allexport;
    unset ERL_AFLAGS;
    unset REBAR_CACHE_DIR;
    unset REBAR_PLT_DIR;
    MANPATH="$(echo "$MANPATH" | sed -r "s%/Users/user/kerl/25.2/lib/erlang/man:/Users/user/kerl/25.2/man:?%%")";
    PATH="$(echo "$PATH" | sed -e "s%/Users/user/kerl/25.2/lib/erl_interface-5.3/bin:%%")";
    PATH="$(echo "$PATH" | sed -e "s%/Users/user/kerl/25.2/bin:%%")";
    unset _KERL_ACTIVE_DIR;
    if [ -n /bin/bash ] || [ -n ]; then
        hash -r;
    fi;
    unset -f kerl_deactivate;
    set +o allexport
}
```
This code may vary depending on what was your environment before calling activate.sh.

**Comment by x0id** [comment]
What I like in this approach (to have all stuff in the closure) is that if you

1. Activate some Erlang installation
2. Upgrade kerl forgetting to de-activate active kerl environment
3. Activate alternative Erlang installation

It will not mess things around, even if the kerl internals change significantly, because the `kerl_deactivate` functions "knows" what to do without a need to consult internal environment variables, which may, for example renamed in the new kerl version.

**Comment by x0id** [comment]
I would also have a tough time reviewing this PR based on the diff output.
Instead, keeping in mind my explanation, simply look at the new code, you will find it straightforward.

**Comment by x0id** [comment]
While I demonstrated the new deactivation code, I found a minor flaw, which I have just fixed in the recent commit. New deactivation code looks better:
```
$ ./kerl emit-activate 25.2 25.2-build /Users/user/kerl/25.2 >/tmp/activate.sh
$ . /tmp/activate.sh
$ type -a kerl_deactivate
kerl_deactivate is a function
kerl_deactivate () 
{ 
    set -o allexport;
    unset ERL_AFLAGS;
    unset REBAR_CACHE_DIR;
    unset REBAR_PLT_DIR;
    MANPATH="$(echo "$MANPATH" | sed -r "s%/Users/user/kerl/25.2/lib/erlang/man:/Users/user/kerl/25.2/man:?%%")";
    PATH="$(echo "$PATH" | sed -e "s%/Users/user/kerl/25.2/lib/erl_interface-5.3/bin:%%")";
    PATH="$(echo "$PATH" | sed -e "s%/Users/user/kerl/25.2/bin:%%")";
    hash -r;
    unset _KERL_ACTIVE_DIR;
    unset -f kerl_deactivate;
    set +o allexport
}
```

**Comment by paulo-ferraz-oliveira** [comment]
I kinda gotta the "change approach" on the first read, but if e.g. one commit changes one bit and then another and then another, it'd be much easier to follow. The way it is (and I'll review it as such) I have to make sure than when deleting old code/creating new code you didn't forget anything (since I'm aware that the tests don't fully cover all functionalities)...

**Comment by paulo-ferraz-oliveira** [comment]
Also, is there no easier way to "prefix" the code execution (e.g. with a function) instead of constantly writing `\$_KERL_CLEANUP` at the end of the "strings"?

File: kerl [hunk 1]
```diff
@@ -1292,52 +1228,119 @@ fi
...
+if [ -n "\${REBAR_CACHE_DIR+x}" ]; then
+    _KERL_CLEANUP="REBAR_CACHE_DIR=\"\$REBAR_CACHE_DIR\"; \$_KERL_CLEANUP"
+else
+    _KERL_CLEANUP="unset REBAR_CACHE_DIR; \$_KERL_CLEANUP"
 fi
+REBAR_CACHE_DIR="$directory/.cache/rebar3"
+
 # https://twitter.com/mononcqc/status/877544929496629248
-_KERL_ERL_AFLAGS_SET="\${ERL_AFLAGS+x}"
 kernel_history=\$(echo "\$ERL_AFLAGS" | \\grep 'kernel shell_history' || true)
```
Comment (paulo-ferraz-oliveira) [inline kerl:1295]: For consistency, we should probably also rename this `_KERL_KERNEL_HISTORY`.

File: kerl [hunk 2]
```diff
@@ -1210,73 +1210,9 @@ emit_activate() {
...
-    unset KERL_ENABLE_PROMPT
-    unset KERL_PROMPT_FORMAT
-    set +o allexport
-}
-kerl_deactivate nondestructive
+if type kerl_deactivate 2>/dev/null | \grep -qw function; then
+    kerl_deactivate
+fi
 
 _KERL=$(command -v kerl)
```
Comment (paulo-ferraz-oliveira) [inline kerl:1218]: We should remove white space -only lines from here until `unset _KERL_VERSION _KERL`. This way we "identify" stuff as blocks, hopefully increasing readability.

File: kerl [hunk 3]
```diff
@@ -1292,52 +1228,119 @@ fi
...
 
 set -o allexport
-_KERL_SAVED_REBAR_CACHE_DIR="\$REBAR_CACHE_DIR"
-_KERL_REBAR_CACHE_DIR_SET="\${REBAR_CACHE_DIR+x}"
-_KERL_SAVED_REBAR_CACHE_DIR="\$REBAR_CACHE_DIR"
-_KERL_REBAR_PLT_DIR_SET="\${REBAR_PLT_DIR+x}"
-_KERL_SAVED_REBAR_PLT_DIR="\$REBAR_PLT_DIR"
+
+_KERL_ACTIVE_DIR="$directory"
+
```
Comment (paulo-ferraz-oliveira) [inline kerl:1233]:
```suggestion
```

File: kerl [hunk 4]
```diff
@@ -1292,52 +1228,119 @@ fi
...
+_KERL_CLEANUP="unset _KERL_ACTIVE_DIR"
+
```
Comment (paulo-ferraz-oliveira) [inline kerl:1235]:
```suggestion
```

File: kerl [hunk 5]
```diff
@@ -1292,52 +1228,119 @@ fi
...
-        _KERL_SAVED_ERL_AFLAGS="\$ERL_AFLAGS"
         ERL_AFLAGS="-kernel shell_history enabled \$ERL_AFLAGS"
     else
         ERL_AFLAGS="-kernel shell_history enabled"
     fi
 fi
 unset kernel_history
+
 # shellcheck source=/dev/null
 if [ -f "$KERL_CONFIG" ]; then . "$KERL_CONFIG"; fi
```
Comment (paulo-ferraz-oliveira) [inline kerl:1311]:
```suggestion
if [ -f "$KERL_CONFIG" ]; then
    . "$KERL_CONFIG"
fi
```

File: kerl [hunk 6]
```diff
@@ -1292,52 +1228,119 @@ fi
...
+    fi
     if [ -n "\$KERL_PROMPT_FORMAT" ]; then
         _KERL_PROMPT_FORMAT="\$KERL_PROMPT_FORMAT"
     else
         _KERL_PROMPT_FORMAT="(%BUILDNAME%)"
     fi
     _KERL_PRMPT=\$(echo "\$_KERL_PROMPT_FORMAT" | sed 's^%RELEASE%^$release^;s^%BUILDNAME%^$build_name^')
     PS1="\$_KERL_PRMPT\$PS1"
+    unset KERL_ENABLE_PROMPT KERL_PROMPT_FORMAT
     unset _KERL_PRMPT _KERL_PROMPT_FORMAT
```
Comment (paulo-ferraz-oliveira) [inline kerl:1326]:
```suggestion
    unset KERL_ENABLE_PROMPT
    unset KERL_PROMPT_FORMAT
    unset _KERL_PROMPT_FORMAT
    unset _KERL_PRMPT
```

**Comment by x0id** [comment]
Sorry, I'm not exactly following your rules for whitespacing.


> Also, is there no easier way to "prefix" the code execution (e.g. with a function) instead of constantly writing `\$_KERL_CLEANUP` at the end of the "strings"?
> 
> **Edit**: e.g. something similar to
> 
> ```shell
> add_cleanup() {
>   CLEANUP="$1; ${CLEANUP}"
> }
> 
> add_cleanup "this"
> add_cleanup "that"
> ```

Not sure this will increase readability. At least I can't see simple solution for the following code:
```
if [ -n "\$_KERL_ERL_CALL_REMOVABLE" ]; then
    PATH="\${_KERL_ERL_CALL_REMOVABLE}:\$PATH"
    _KERL_CLEANUP="
        PATH=\"\\\$(echo \"\\\$PATH\" | sed -e \"s%\$_KERL_ERL_CALL_REMOVABLE:%%\")\"
        \$_KERL_CLEANUP
    "
fi
```

**Comment by paulo-ferraz-oliveira** [comment]
> Sorry, I'm not exactly following your rules for whitespacing.

If possible, act on the suggestions first and what comments remain I'll try to be more explicit. 😄

**Comment by paulo-ferraz-oliveira** [comment]
Wouldn't this work?

```bash
  add_cleanup() {
    _KERL_CLEANUP="
      $1
      $_KERL_CLEANUP
    "
  }
```

Probably I'm not thinking about some detail because you're in the context of a generated file, but I believe it would at least increase the chance of not forgetting to suffix the command with the cleanup "action".

File: kerl (hunk 2 above)
Comment (paulo-ferraz-oliveira) [reply kerl:1218]: i.e.

```bash
_KERL=$(command -v kerl)

if [ -n "\$_KERL" ]; then
  _KERL_VERSION=\$(\$_KERL version)
fi

if [ -n "\$_KERL_VERSION" ] && [ "\$_KERL_VERSION" != "$KERL_VERSION" ]; then
  echo "WARNING: this Erlang/OTP installation appears to be stale. Please consider reinstalling."
  echo "         It was created with kerl $KERL_VERSION, and the current version is \$_KERL_VERSION."
fi

unset _KERL_VERSION _KERL
```

becomes

```bash
_KERL=$(command -v kerl)
if [ -n "\$_KERL" ]; then
  _KERL_VERSION=\$(\$_KERL version)
fi
if [ -n "\$_KERL_VERSION" ] && [ "\$_KERL_VERSION" != "$KERL_VERSION" ]; then
  echo "WARNING: this Erlang/OTP installation appears to be stale. Please consider reinstalling."
  echo "         It was created with kerl $KERL_VERSION, and the current version is \$_KERL_VERSION."
fi
unset _KERL_VERSION _KERL
```

but I can't suggest this via the GH interface 😢

File: kerl [hunk 7]
```diff
@@ -1292,52 +1228,119 @@ fi
 unset _KERL_VERSION _KERL
```
Comment (paulo-ferraz-oliveira) [inline kerl:1228]:
```suggestion
unset _KERL
unset _KERL_VERSION
```

File: kerl (hunk 1 above)
Comment (x0id) [reply kerl]: @paulo-ferraz-oliveira I addressed most if not any of the concerns, the `add_cleanup` is really helpful, thank you.
Please let me know if anything else needs to be distilled.

**Comment by x0id** [comment]
@paulo-ferraz-oliveira I addressed most if not any of the concerns, the `add_cleanup` is really helpful, thank you.
Please let me know if anything else needs to be distilled.

**Comment by paulo-ferraz-oliveira** [comment]
I'll check it, @x0id, thanks.

**Comment by paulo-ferraz-oliveira** [comment]
I'll approve this, but I'll also invite @jadeallenx to review. Thanks.

**Review by paulo-ferraz-oliveira — APPROVED** [review]

**Review by jadeallenx — APPROVED** [review]
:shipit:

**Comment by paulo-ferraz-oliveira** [comment]
@x0id, this is squash-merged. Let us know if you need a tag, otherwise this can wait for a further update.

**Comment by x0id** [comment]
No urgency with tagging. I will try to add tests for csh activate/deactivate soon. Maybe then it will make more sense to release.
Thank you.

**Comment by jadeallenx** [comment]
This is all great stuff - thanks!
